### DISCUSSION TITLE
Generic Return Value for Moving Averages?

### DISCUSSION THREAD

**Discussion by cperret977** [DaveSkender/Stock.Indicators#1061] [Help and support]
First things first: I'm trying to cram learning C# programming and the use of this excellent library into the same project, so please... be gentle with me.

What I'm trying to do: I want to create a function that I can pass an integer value to that is used to select one of several moving averages, then return the results of the moving average to the calling function where those results will be used to determine trades. I'd like to make this function so that I can use the same selection parameters in multiple places and always get the same MA for the same input, without having to repeat a bunch of code.

Example:
```
public static void System1( [other parameters], int intMAtype)
{
   [do some stuff]
   [some type] MA;
   MA = GetMovingAverage(quotes, candlePart, intMAtype, intMAlength);
   [do more stuff using the MA]
}

public static [some return type] GetMovingAverage(Quotes quotes, CandlePart candlePart, int intMAtype, int intMAlength)
{
   [some return type] results;
   switch (intMAtype)
   {
      case 1:
         results = quotes.Use(candlePart).GetSMA(intMAlength);
         break;
      case 2:
         results = [do exponential moving average]
         break;
      [continue for other types of moving averages]
   }
   return results;
}
```
So, I'm trying to figure out the proper type for MA in the System1 function and proper return type for the GetMovingAverage() function (and, by extension, the type for results in that function) - or find out that it's not possible to do what I'm trying to do and there's something else I should be doing instead. Is it just a matter of making all of these types an IResult or IReusableResult? Those seem to be what the separate MA result types are derived from, but I can't tell if that will get me what I need (date and MA value - same as one of the specific results).

**Comment by DaveSkender** [comment]
You can definitely do this.  A while back, I’d done something like this for [Rolling Moving Average](https://github.com/DaveSkender/Stock.Indicators/blob/c0597b1944f91a7212089c3288ce838dfb9e9df4/src/m-r/MaEnvelopes/MaEnvelopes.Series.cs), but it might benefit from a bit of refactoring to simplify.  It might give you some inspiration.

**Reply by DaveSkender** [reply]
If you look at the [EMA results class](https://github.com/DaveSkender/Stock.Indicators/blob/c0597b1944f91a7212089c3288ce838dfb9e9df4/src/e-k/Ema/Ema.Models.cs) you can see how we’re using `IReusableResult` to create a common alias called `Value`. You’d really only need it in your results model to further enable chaining options.

The benefits of using `ResultBase` or `IResult` is less about the basic form and more about getting all the [results utilities](https://dotnet.stockindicators.dev/utilities/#utilities-for-indicator-results) that come with it.  `ResultBase` already defines the `Date` property, so that’s why I’m only mapping it.

You could do something similar to catch your results, like:

```csharp
public class MaResult : ResultBase, IReusableResult
{
    public MaResult(DateTime date)
    {
        Date = date;
    }
    public double? Ma { get; set; }
    double? IReusableResult.Value => Ma; // recommended for chaining
}
```

**Reply by DaveSkender** [reply]
Also, take a look at the [MaType enum](https://github.com/DaveSkender/Stock.Indicators/blob/c0597b1944f91a7212089c3288ce838dfb9e9df4/src/_common/Enums.cs#L38-L51).  It may save you from redefining with `int` values separately.  In my RMA example, you can see how I use this in a `switch` map similarly to what you’re trying to do.

**Reply by DaveSkender** [reply]
Under your switch, it’d look something like:

```csharp
case MaType.ALMA:

results = quotes
  .Use(candlePart)
  .GetAlma(lookbackPeriods)
  .Select(x => new MaResult(x.Date)
    {
      Ma = x.Alma
      // or Ma = x.Value, if you want to use the alias
    });
break;
```

Your results would return as `IEnumerable<MaResult> `
